manim/utils/units.py

```diff
@@ -0,0 +1,86 @@
+import copy
+import logging
+from functools import wraps
+
+import numpy as np
+
+from ..config import config
+
+
+def _size_from_dimension(dim=0, pixels=False):
+    if dim == 0:
+        return config["frame_width"] if not pixels else config["pixel_width"]
+    elif dim == 1:
+        return config["frame_height"] if not pixels else config["pixel_height"]
+    else:
+        # TODO: determine size in z-direction
+        logging.getLogger("manim").warning(
+            f"Z-direction is not supported for units - returning size of x-dimension."
+        )
+        return config["frame_width"] if not pixels else config["pixel_width"]
+
+
+def _is_dimensional(value):
+    return isinstance(value, np.ndarray) and value.shape[0] <= 3
+
+
+class Unit:
+    def __init__(self, value):
+        self._value = value
+        self._converter = lambda val, dim: val
+
+    @property
+    def value(self):
+        return copy.deepcopy(self._value)
+
+    @property
+    def converter(self):
+        return self._converter
+
+    def convert(self, dim=None):
+        if dim:
+            return self.converter(self.value, dim=dim)
+        elif _is_dimensional(self.value):
+            return_value = self.value
+
+            for i in range(0, return_value.shape[0]):
+                return_value[i] = self.converter(return_value[i], i)
+            return return_value
+        else:
+            raise ValueError("Unable to determine dimension.")
+
+
+class MUnit(Unit):
+    def __init__(self, value):
+        super(MUnit, self).__init__(value)
+
+
+class PixelUnit(Unit):
+    def __init__(self, value):
+        super(PixelUnit, self).__init__(value)
+        self._converter = (
+            lambda val, dim: val
+            / _size_from_dimension(dim, True)
+            * _size_from_dimension(dim)
+        )
+
+
+class PercentUnit(Unit):
+    def __init__(self, value):
+        super(PercentUnit, self).__init__(value)
+        self._converter = lambda val, dim: val / 100 * _size_from_dimension(dim)
+
+
+def accepts_unit(f, dim=None):
+    @wraps(f)
+    def wrapper(*args, **kwargs):
+        # TODO figure what to do when dim=None, preferable determine the dim.
+        return f(
+            *[v.convert(dim) if isinstance(v, Unit) else v for v in args],
```

Comment:
Why not making Unit class accept one unit argument (so than `Unit(Unit(Unit(v))) == Unit(v)`. You could then use `Unit(v).convert(dim)` without the if clause. I also think it will be useful in the future, as a quick way to make sure a "value" is really an Unit object

Reply:
I think we shouldn't get rid of the if clause. Right now, this calls `convert` for each Unit object it sees. If we get rid of the if clause, we will have to i) create a new Unit object, and b) call its `convert` method *for each argument in each function decorated with this.* That will surely create a lot of overhead. Function calls in python are sadly not cheap.